### PR TITLE
feat: Implement ScrollViewer.ViewChanging

### LINKED ISSUE #78: ScrollViewer.ViewChanging doesn't fire on android
<ScrollViewer VerticalScrollMode="Enabled" HorizontalScrollMode="Enabled"
        ViewChanged="ScrollViewer_ViewChanged" ViewChanging="ScrollViewer_ViewChanging">
        <Border Height="1000" Width="1000" Background="Red"/>
    </ScrollViewer>

The ScrollViewer_ViewChanged fires, the ScrollViewer_ViewChanging doesn't. 
Tried with 1.31.0-dev.91
[Repo inside](https://github.com/Henk7/Uno.QuickStart/commits/scrollviewer)

### PR DESCRIPTION
**Pull request by Youssef1313**
*PR is still early to be ready. Opening as a draft for now*

GitHub Issue (If applicable): closes #78

## PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

ViewChanging is not implemented.

## What is the new behavior?

It's now implemented.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/ScrollViewer.cs [hunk 1]
```diff
@@ -769,22 +769,7 @@ public static void SetIsDeferredScrollingEnabled( global::Windows.UI.Xaml.Depend
...
-				global::Windows.Foundation.Metadata.ApiInformation.TryRaiseNotImplemented("Windows.UI.Xaml.Controls.ScrollViewer", "event EventHandler<ScrollViewerViewChangingEventArgs> ScrollViewer.ViewChanging");
-			}
-			[global::Uno.NotImplemented("__ANDROID__", "__IOS__", "NET461", "__WASM__", "__SKIA__", "__NETSTD_REFERENCE__", "__MACOS__")]
-			remove
-			{
-				global::Windows.Foundation.Metadata.ApiInformation.TryRaiseNotImplemented("Windows.UI.Xaml.Controls.ScrollViewer", "event EventHandler<ScrollViewerViewChangingEventArgs> ScrollViewer.ViewChanging");
-			}
-		}
-		#endif
+		// Skipping already declared event Windows.UI.Xaml.Controls.ScrollViewer.ViewChanging
```
Comment (Youssef1313) [inline src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/ScrollViewer.cs:772]: Change here is made manually. I don't know how to run the generator that generates this file? Building any of the solution filters doesn't seem to do that.

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs [hunk 2]
```diff
@@ -1298,6 +1303,7 @@ private void RequestUpdate()
 
 		private void Update(bool isIntermediate)
 		{
+			ViewChanging?.Invoke(this, new ScrollViewerViewChangingEventArgs(true, null, null)); // TODO: Fix values.
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: This is the part where I'm struggling.

From UWP docs:

> `IsInertial` is **true** for manipulations that occur:
>
> - using a mouse wheel.
> - using key strokes on `ScrollViewer` and other controls that support direct manipulation.
> - by calling `ChangeView` with animation enabled.
>
> `IsInertial` is **false** for manipulations that occur:
>
> - by dragging the Thumb on a scrollbar.
> - using key strokes on controls that don't support direct manipulation.
> - by calling ChangeView with animation disabled.

What's the equivalent of `mouse wheel` vs `dragging` in mobile apps? How to access this information.

---

I don't get what `FinalView` and `NextView` are from the docs:

https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.scrollviewerviewchangingeventargs.finalview?view=winrt-20348

https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.scrollviewerviewchangingeventargs.nextview?view=winrt-20348

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs [hunk 3]
```diff
@@ -1310,7 +1316,6 @@ private void Update(bool isIntermediate)
 			ScrollOffsets = new Point(_pendingHorizontalOffset, _pendingVerticalOffset);
 			InvalidateViewport();
 #endif
-
 			ViewChanged?.Invoke(this, new ScrollViewerViewChangedEventArgs { IsIntermediate = isIntermediate });
```
Comment (Youssef1313) [inline src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1319]: fwiw, this doesn't match the UWP API. The property shouldn't have a public setter. This is unrelated to the PR. Let me know if you want a separate tracking issue for that.

File: src/Uno.UI/Uno.UI.csproj [hunk 4]
```diff
@@ -174,6 +174,10 @@
 		</ProjectReference>
 	</ItemGroup>
 
+	<ItemGroup>
+	  <UpToDateCheckInput Remove="UI\Xaml\Controls\ScrollViewer\ScrollViewerViewChangingEventArgs.cs" />
+	</ItemGroup>
```
Comment (Youssef1313) [inline src/Uno.UI/Uno.UI.csproj:179]: This was auto-generated by Visual Studio. I'm not sure if this is needed.

**Review by Youssef1313 — COMMENTED** [review]
cc @jeromelaban @MartinZikmund for help here. Thanks in advance!

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs (hunk 2 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: @dr1rrb will probably know the answer 🙂

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs (hunk 3 above)
Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1319]: Which property please?

File: src/Uno.UI/Uno.UI.csproj (hunk 4 above)
Comment (MartinZikmund) [reply src/Uno.UI/Uno.UI.csproj:179]: Yes, VS often generates these, you can undo this change

File: src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/ScrollViewer.cs (hunk 1 above)
Comment (MartinZikmund) [reply src/Uno.UI/Generated/3.0.0.0/Windows.UI.Xaml.Controls/ScrollViewer.cs:772]: There is UWPSyncGenerator which can do this, but we usually don't run it for single changes like this, it is usually enough to adjust the #if and `[NotImplemented]` accordingly

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs (hunk 3 above)
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1319]: IsIntermediate

Comment (MartinZikmund) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1319]: I see, you can make the setter `internal` then, as it will match UWP. It is a breaking change though, so you will need to update `PackageDiffIgnore.xml` accordingly. You can for now just make it internal and commit the change, and the build will tell you specifically how to update that file.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1319]: I think this might be suitable for a separate PR since this is not part of `ViewChanging`. Going to open another issue

File: src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs (hunk 2 above)
Comment (davidjohnoliver) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: btw if you want to test UWP touch behavior, deploying the UWP head to 'Simulator' instead of 'Local Machine' is useful

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: > What's the equivalent of `mouse wheel` vs `dragging` in mobile apps? How to access this information.

There is no real equivalent to _mouse wheel_ on mobile apps, unless you connect a mouse to your device (on Android I mean ... mouses are poorly handled like a touch pointer on iOS, at OS level).

I would put the `IsInertial` in parallel of the [`ViewChanged.IsIntermediate`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.scrollviewerviewchangedeventargs.isintermediate?view=winrt-20348#Windows_UI_Xaml_Controls_ScrollViewerViewChangedEventArgs_IsIntermediate) which only indicates if the event is the final one or not, so we can avoid to do useless work and preserve resources for a more performant scrolling experience.

From my understanding, the `IsInertial` means that the user interaction is already over (key up / wheel events are already over), so we already know where the scrolling is going to stop, but the `ScrollViewer` is going to animate that change. On the other side, when dragging (the content -touch/pen- or the scrollbars -mouse/pen-), we don't know yet where the scrolling is going to stop. The application can then adjust its logic to get a more responsive scrolling (like triggering the loading of the element that is going to be visible). I would however confirm that with touch we do have `IsInertial==false` when dragging and when flicking, it's only an assumption.

And about where to get the info: Well you are not at the right place here, the view has already changed, and the info has already been lost. You have to raise that event at each place which can trigger the scrolling.
* For programmatic scrolling, you can put it right before the [`ChangeViewNative`](https://github.com/unoplatform/uno/blob/master/src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs#L1385). The `disableAnimation` flag is already accessible here.
* For **MANAGED** `ScrollBar` (WASM, Skia, macOS) I would assume it will be already correct as the end in the `ChangeViewCore` (FYI cf. [`On<Vertical|Horizontal>ScrollBarScrolled`](https://github.com/unoplatform/uno/blob/4029e71b5da907ca53fe7e85d0590b5ca533b786/src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs#L1158)) 
* For **MANAGED** `ScrollContentPresenter` (Skia, macOS) you will have to look at the [`ScrollContentPresenter_PointerWheelChanged`](https://github.com/unoplatform/uno/blob/4029e71b5da907ca53fe7e85d0590b5ca533b786/src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.Managed.cs#L212) and [`BeginTouchScroll`](https://github.com/unoplatform/uno/blob/4029e71b5da907ca53fe7e85d0590b5ca533b786/src/Uno.UI/UI/Xaml/Controls/ScrollContentPresenter/ScrollContentPresenter.Managed.cs#L252)
* And the trickiest part, for **NATIVE** `ScrollContentPresenter` (WASM, iOS, Android) you will have to register to native event of each platform:
    * iOS: https://developer.apple.com/documentation/uikit/uiscrollviewdelegate/1619394-scrollviewwillbegindragging but I assume that you won't be able to make a distinction between input kinds.
    * Android: 
        * keyboard: https://github.com/unoplatform/uno/blob/4029e71b5da907ca53fe7e85d0590b5ca533b786/src/Uno.UI.BindingHelper.Android/Uno/UI/UnoTwoDScrollView.java#L318
        * touch: https://github.com/unoplatform/uno/blob/4029e71b5da907ca53fe7e85d0590b5ca533b786/src/Uno.UI.BindingHelper.Android/Uno/UI/UnoTwoDScrollView.java#L437
        * wheel: ???
     * WASM: ??? (I see only the `scroll` event, but I think that it's raised only after the scroll has began, and browsers have started to enable async scrolling so it will become less and less relevant)

Even if not equivalent to UWP, I think that we could accept to raise the `ViewChanging` right before the `ViewChanged`, no matter id the UI has already been updated, but I would try to do that only where we are not able to get it at the right time, and in that case I think we won't be able to provide the right info.

> 
> I don't get what `FinalView` and `NextView` are from the docs:
> 

From my understanding the `NextView` is the view which will be visible (at the top/center/bottom of the viewport?) once the scroll has occured and the `ViewChangedEvent` corresponding to that `ViewChanging` event has been raised, no matter if the scrolling is still pending (i.e. while dragging) or animating (i.e. inertia). The `FinalView` is the view that is going to be visible once scrolling has stopped, including inertia. So I assume that this will stay `null` while dragging as we cannot yet know where the scroll will end, but it will be full-filled while animating as we already know where the scroll will stop.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: Thanks a lot @dr1rrb. This is very useful. Couple of points:


- In UWP, Looks like `FinalView` is not null when `IsInertial == true`

    image

    Seems it's always going to be equals to `NextView` (at least based on what I've seen so far).

- Regarding the value of `IsInertial`, it seems unrelated to `IsIntermediate` (`IsIntermediate` can be true for both mouse wheel and scrollbar - while `IsInertial` is true for mousewheel but false for scrollbar)

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: Also from my experiments with UWP, it looks like ViewChanging can be fired multiple times (while animating) before ViewChanged is fired. ie, Not every ViewChanging is followed by ViewChanged, only the final one. The code path in `ChangeViewCore` and `Update` doesn't seem to agree with that behavior (tested on Wasm).

How far can we diverge from UWP behavior?

Comment (dr1rrb) [reply src/Uno.UI/UI/Xaml/Controls/ScrollViewer/ScrollViewer.cs:1306]: > 
> 
> Thanks a lot @dr1rrb. This is very useful. Couple of points:
> 
>     * In UWP, Looks like `FinalView` is not null when `IsInertial == true`
>       Seems it's always going to be equals to `NextView` (at least based on what I've seen so far).

You should try to put a list of elements that are just a bit slower than the viewport, then request programmatically a `ChangeView` to the last one with animations enabled. I'm pretty sure that the `NextView` won't be directly the `FinalView`. 
> 
>     * Regarding the value of `IsInertial`, it seems unrelated to `IsIntermediate` (`IsIntermediate` can be true for both mouse wheel and scrollbar - while `IsInertial` is true for mousewheel but false for scrollbar)

Yeah by "comparable", I mean in term of usage: both are used to indicate into which step the `ScrollViewer` is, so applications can enhance their behavior.

> Also from my experiments with UWP, it looks like ViewChanging can be fired multiple times (while animating) before ViewChanged is fired. ie, Not every ViewChanging is followed by ViewChanged, only the final one. 

The `ScrollViewer` is a really weird control, and is highly optimized. The `ViewChanged` (and most probably the `ViewChanging`) is raised in a "best effort" way: it does not wait for your handler to continue / complete scrolling. It will raise as much event as it can, and as far as I know it does raise multiple `ViewChanged` on UWP also. You should use logging, and not breakpoints (i would not even recommend tracepoint as they are pretty slow).
